### PR TITLE
Converts all editor3 functionalities to use redux [SDESK-594]

### PR DESCRIPTION
**Pull request by gbbr**
Additionally:

- Fixes an issue where links with typos were not styled accordingly ([SDESK-608](https://dev.sourcefabric.org/browse/SDESK-608))
- Introduces mocking functionality into the `ng` service to ease testing components and other aspects of the redux architecture that make use of it
- Adds some basic tests (more to come in follow-up PR)

### CODE REVIEW

File: scripts/core/editor3/reducers/index.jsx [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
-// custom combine of reducers because the data is needed in all reucers
-const editorReducers = (state, action) => {
-    var newState = spellchecker(state, action);
+// Returns a new reducer which chains the state and action throught the given
+// list of reducers.
+const chainReduce = (...reducers) => (state = {}, action) => {
+    let newState = {...state};
 
-    return editor3(newState, action);
+    reducers.forEach((r) => {
```
Comment (petrjasek) [inline scripts/core/editor3/reducers/index.jsx]: this seems like something for reduce: `return reducers.reduce((state, r) => r(state, action), newState);`

File: scripts/core/editor3/tests/Editor3Component.spec.jsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    });
+});
+
+describe('Editor3.createEditorStore', () => {
+    it('should initialize with correct values', inject(($q) => {
+        const setLanguage = jasmine.createSpy(),
+            getDict = jasmine.createSpy().and.callFake(() => $q.when(null)),
+            isCorrectWord = jasmine.createSpy();
+
+        ng.mock('spellcheck', {setLanguage, getDict, isCorrectWord});
```
Comment (petrjasek) [inline scripts/core/editor3/tests/Editor3Component.spec.jsx]: given that it's based on `$injector`, you could use probably:
```
beforeEach(inject(($provide) => {
    $provide.service('spellchecker', ($q) => ({
        setLanguage: jasmine.createSpy(),
        getDict: jasmine.createSpy().and.returnValue($q.when(null)),
        isCorrectWord: jasmine.createSpy()
    }));
}));
```
so you wouldn't need to add that mock func to `ng` and you don't have to clean it after

File: scripts/core/editor3/reducers/index.jsx (hunk 1 above)
Comment (gbbr) [inline scripts/core/editor3/reducers/index.jsx]: *doh* 😄

**Comment by gbbr** [comment]
@petrjasek nice tips, I've updated the PR with your suggestions

File: scripts/core/editor3/components/toolbar/BlockStyleControls.jsx [hunk 3]
```diff
@@ -19,53 +20,45 @@ const BLOCK_TYPES_STYLE = {
...
+};
 
-    render() {
-        const {editorState, options} = this.props;
-        const selection = editorState.getSelection();
-        const blockType = editorState
-            .getCurrentContent()
-            .getBlockForKey(selection.getStartKey())
-            .getType();
+const mapStateToProps = (state, ownProps) => ({
```
Comment (ioanpocol) [inline scripts/core/editor3/components/toolbar/BlockStyleControls.jsx]: You can delete the parameter ownProps if you don't use it.

File: scripts/core/editor3/components/toolbar/InlineStyleControls.jsx [hunk 4]
```diff
@@ -9,47 +10,45 @@ const INLINE_STYLES = {
...
+};
+
+InlineStyleControlsComponent.propTypes = {
     editorState: React.PropTypes.object,
     options: React.PropTypes.array,
-    onChange: React.PropTypes.func.isRequired
+    toggleInlineStyle: React.PropTypes.func
 };
+
+const mapStateToProps = (state, ownProps) => ({
```
Comment (ioanpocol) [inline scripts/core/editor3/components/toolbar/InlineStyleControls.jsx]: Also here can be deleted ownProps

File: scripts/core/editor3/components/toolbar/link-button/LinkControl.jsx [hunk 5]
```diff
@@ -182,8 +131,23 @@ function LinkStrategy(contentBlock, callback) {
...
-LinkControl.propTypes = {
+LinkControlComponent.propTypes = {
     editorState: React.PropTypes.object.isRequired,
     editorRect: React.PropTypes.object.isRequired,
-    onChange: React.PropTypes.func.isRequired
+    applyLink: React.PropTypes.func.isRequired,
+    removeLink: React.PropTypes.func.isRequired
 };
+
+const mapStateToProps = (state, ownProps) => ({
```
Comment (ioanpocol) [inline scripts/core/editor3/components/toolbar/link-button/LinkControl.jsx]: The same here

**Comment by gbbr** [comment]
@ioanpocol done

**Pull request merged**